sgl-kernel/csrc/cpu/gemm_int4_w4a8.cpp

```diff
@@ -0,0 +1,617 @@
+  const float* bias_ptr = bias.has_value() ? bias.value().data_ptr<float>() : nullptr;
+
+  at::parallel_for(0, num_blocks, 1, [&](int64_t begin, int64_t end) {
+    for (const auto i : c10::irange(begin, end)) {
+      int64_t mc = parallel_on_M ? i / Nc : 0;
+      int64_t nc = parallel_on_M ? i % Nc : i;
+      int64_t mc_end = parallel_on_M ? mc + 1 : Mc;
+
+      for (int mci = mc; mci < mc_end; ++mci) {
+        int64_t m_size = mci * block_m + block_m > M ? M - mci * block_m : block_m;
+        alignas(64) float y_buf[m_size][BLOCK_N];
+        // copy bias to y_buf if bias is not None
+        auto bias_data = bias_ptr ? bias_ptr + nc * BLOCK_N : nullptr;
+        copy_bias<BLOCK_N>(bias_data, y_buf[0], m_size);
+        for (int kci = 0; kci < Kc; ++kci) {
+          int32_t* compensation_ptr =
+              sym_quant_a ? nullptr
+                          : (int32_t*)(void*)(b_ptr + (nc * Kc + kci) * (BLOCK_N * (BLOCK_K / 2 + sizeof(int32_t))) +
+                                              BLOCK_K * BLOCK_N / 2) /*Bcomp*/;
+          _dequant_gemm_accum<BLOCK_N, BLOCK_N / 2, sym_quant_a>(
+              y_buf[0] /*C*/,
+              (uint8_t*)a_ptr + mci * block_m * K + kci * BLOCK_K /*A*/,
+              a_scales_ptr + mci * block_m /*scales_a*/,
+              a_qzeros_ptr + mci * block_m /*qzeros_a*/,
+              b_ptr + (nc * Kc + kci) * (BLOCK_N * (BLOCK_K / 2 + sizeof(int32_t))),
+              b_scales_ptr + nc * BLOCK_N * num_groups + kci / block_per_group * BLOCK_N /*scales_b*/,
+              b_qzeros_ptr + nc * BLOCK_N * num_groups + kci / block_per_group * BLOCK_N /*qzeros_b*/,
+              compensation_ptr /*Bcomp*/,
+              m_size /*M*/,
+              BLOCK_K /*K*/,
+              K /*lda*/,
+              BLOCK_N /*ldc*/,
+              use_brgemm);
+        }
+        // store y_buf to output with dtype conversion
+        store_out<out_dtype, BLOCK_N>(y_buf[0], c_ptr + mci * block_m * N + nc * BLOCK_N, m_size, N /*lda*/);
+      }
+    }
+    if (use_brgemm) {
+      at::native::cpublas::brgemm_release();
+    }
+  });
+}
+
+}  // anonymous namespace
+
+at::Tensor int4_scaled_mm_cpu_with_quant(
+    const at::Tensor& input,
+    const at::Tensor& weight,
+    const at::Tensor& weight_scales,
+    const at::Tensor& weight_qzeros,
+    const std::optional<at::Tensor>& bias,
+    at::ScalarType output_dtype) {
+  RECORD_FUNCTION("sgl-kernel::int4_scaled_mm_cpu_with_quant", std::vector<c10::IValue>({input, weight}));
+
+  int64_t M_a = input.size(0);
+  int64_t K_a = input.size(1);
+  int64_t lda = input.stride(0);
+
+  const auto st = input.scalar_type();
+  TORCH_CHECK(
+      st == at::kBFloat16 || st == at::kHalf, "int4_scaled_mm_cpu_with_quant: expect A to be bfloat16 or half.");
+
+  auto Aq = at::empty({M_a, K_a}, input.options().dtype(c10::kByte));
+  auto As = at::empty({M_a}, input.options().dtype(at::kFloat));
+  auto Azp = at::ones({M_a}, input.options().dtype(at::kInt)).mul(128);
```

Comment:
the way that `Aq` is created just assume that `sym_quant_a` is false, and A is byte (uint8_t).

if we do support `sym_quant_a` is true, A would be char (int8_t).

to eliminate the ambiguity here, you just allocate a buffer of `Aq + As + Azp` with byte, and then reinterpret_cast in following calls.

Reply:
Sure, have refined here to use a buffer `Aq + As + Azp`